### ISSUE TITLE
json output for snapm CLI

### ISSUE THREAD

**Issue by trgill** [snapshotmanager/snapm#3]
Please consider JSON output for the CLI.  We'd like to base the Ansible role for snapshots on snapm and JSON would make parsing the output easier.

FYI - this is the Ansible role : https://github.com/linux-system-roles/snapshot

**Comment by trgill** [comment]
Maybe a better approach would be for the role to use the python API for snapm?  (rather than parse JSON output)

**Comment by trgill** [comment]
> API access is another option but the snapm APIs aren't too well defined yet - I've been mostly focused on getting basic functionality working so far. If we can have a discussion around what you need in terms of an API that might help to focus this work.

Could we use an API to load your plugins?  At a glance, it looks like you are planning to have a plugin implement an interface to support the different technologies that allow snapshots?

**Comment by bmr-cymru** [comment]
> Could we use an API to load your plugins? At a glance, it looks like you are planning to have a plugin implement an interface to support the different technologies that allow snapshots?

Yes that's right. The `Manager` class in `snapm.manager` is responsible for loading the plugins and managing the state of the snapshot sets that are discovered on the system. It also provides the interface to manipulate snapshot sets and change their state.

You can see how this is used in the `snapm.command` module. The CLI creates an instance of the `Manager` class and uses it to interact with the plugin backend, e.g.:

```python
def create_snapset(manager, name, mount_points):
    """
    Create a new snapshot set from a list of mount points.
    
    :param manager: The manager context to use
    :param name: The name of the new snapshot set
    :param mount_points: A list of mount points to snapshot
    """
    return manager.create_snapshot_set(name, mount_points)

...

def _create_cmd(cmd_args):
    """
    Create snapshot set command handler.
    Attempt to create the specified snapshot set.
    
    :param cmd_args: Command line arguments for the command
    :returns: integer status code returned from ``main()``
    """
    manager = Manager()
    snapset = create_snapset(manager, cmd_args.snapset_name, cmd_args.mount_points)
    _log_info(
        "Created snapset %s with %d snapshots", snapset.name, snapset.nr_snapshots
    )
    print(snapset)
    return 0
```

Nothing's really formalised yet and the interface is subject to change but this would likely be the basis for a future API for snapshot manager.

**Comment by bmr-cymru** [comment]
@trgill we have JSON output available in [0.3.0](https://github.com/snapshotmanager/snapm/releases/tag/v0.3.0) - the Python API is still evolving a bit but how does it look for your uses now?

**Comment by bmr-cymru** [comment]
I think we're settled on using the `Manager` interface now in the snapshot role - closing this as completed for now, but please file new issues for anything that's missing or unclear.
